cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginHandler.java

```diff
@@ -40,17 +152,124 @@
+            final String servicePath = serviceUrl.getPath();
+            if (servicePath.startsWith("/prereg/") || "/prereg".equals(servicePath)) {
+                osfCampaign.setOsfCampaign(
+                        "prereg",
+                        "OSF&nbsp;Preregistration&nbsp;Challenge",
+                        "OSF<br><br>Preregistration<br>Challenge",
+                        "?campaign=prereg"
+                );
+            } else if (servicePath.startsWith("/erpc/") || "/erpc".equals(servicePath)) {
+                osfCampaign.setOsfCampaign(
+                        "erpc",
+                        "Open&nbsp;Science&nbsp;Framework<br><br>Election&nbsp;Research&nbsp;Preacceptance&nbsp;Competition",
+                        "OSF<br><br>Election&nbsp;Research<br>Preacceptance<br>Competition",
+                        "?campaign=erpc"
+                );
+            } else if (servicePath.startsWith("/preprints/") || "/preprints".equals(servicePath)) {
+                osfCampaign.setOsfCampaign(
+                        "osf-preprints",
+                        "OSF&nbsp;Preprints",
+                        "OSF&nbsp;Preprints",
+                        "?campaign=osf-preprints"
+                );
+            } else if ("/login/".equals(servicePath) || "/login".equals(servicePath)) {
+                final String serviceQuery = serviceUrl.getQuery();
+                URL nextUrl = null;
+                if (serviceQuery != null) {
+                    final Map<String, String> queryPairs = parseUrlQuery(serviceQuery);
+                    try {
+                        nextUrl = new URL(queryPairs.get("next"));
+                    } catch (final MalformedURLException e) {
+                        LOGGER.error(String.format("Malformed Service URL: %s", e.toString()));
+                    }
+                    if (nextUrl != null && nextUrl.getPath() != null) {
+                        if (nextUrl.getPath().startsWith("/preprints/") || "/preprints".equals(nextUrl.getPath())) {
+                            osfCampaign.setOsfCampaign(
+                                    "osf-preprints",
+                                    "OSF&nbsp;Preprints",
+                                    "OSF&nbsp;Preprints",
+                                    "?campaign=osf-preprints"
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        if (isInstitutionLogin(context)) {
+            osfCampaign.setInstitutionLogin(true);
+        }
+
+        return osfCampaign;
+    }
+
+    /**
+     * Parse the query string in URL and store the result as key-value pairs in a map.
+     * Support query parameter with empty value. Does not support multiple parameter with the same key.
+     *
+     * @param query the query string
+     * @return a map of query key-value pairs
+     */
+    private Map<String, String> parseUrlQuery(final String query) {
```

Comment:
`java.net.URL` lacks a util method to further parse query part into key-value pairs. This helper method is based on: http://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection. However, I decide not to support multiple keys with the same name for simplicity.